File: design-add-and-search-words-data-structure/design-add-and-search-words-data-structure.java
class TrieNode {
    TrieNode[] children = new TrieNode[26];
    boolean eow = false;
}

class WordDictionary {
    TrieNode root;

    /** Initialize your data structure here. */
    public WordDictionary() {
        this.root = new TrieNode();
    }
    
    public void addWord(String word) {
        TrieNode curr = this.root;
        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);
            int idx = ch - 'a';
            
            if (curr.children[idx] == null) {
                curr.children[idx] = new TrieNode();
            }
            
            curr = curr.children[idx];
        }
        
        curr.eow = true;
    }
    
    public boolean search(String word) {        
        return search(word, 0, root);
    }
    
    private boolean search(String word, int i, TrieNode curr) {
        if (curr == null) {
            return false;
        }
        
        if (i == word.length()) {
            return curr.eow;
        }
        
        char ch = word.charAt(i);
        if (ch == '.') {
            for (TrieNode node : curr.children) {
                if (search(word, i + 1, node)) {
                    return true;
                }
            }
                    
            return false;
        } else {
            int idx = ch - 'a';
            curr = curr.children[idx];
            return search(word, i + 1, curr);
        }
            
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: design-add-and-search-words-data-structure/design-add-and-search-words-data-structure.py
class TrieNode:
    
    def __init__(self):
        self.children = [None] * 26
        self.is_word = False
        
class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        curr = self.root
        for c in word:
            idx = ord(c) - ord('a')
            if curr.children[idx] is None:
                curr.children[idx] = TrieNode()
            curr = curr.children[idx]
        
        curr.is_word = True

    def search(self, word: str) -> bool:
        
        def helper(curr, i):
            if curr is None:
                return False
            
            if i == len(word):
                return curr.is_word
            
            c = word[i]
            if c == '.':
                for child in curr.children:
                    if helper(child, i + 1):
                        return True
                return False
            else:
                idx = ord(c) - ord('a')
                return helper(curr.children[idx], i + 1)
        
        return helper(self.root, 0)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)